rectools/models/rank_torch.py

```diff
@@ -0,0 +1,242 @@
+        For item-item similarity models item similarity vectors are viewed as factors.
+    dtype: Optional[torch.dtype]
+        dtype to convert non-torch tensors to.
+        Don not convert if provided dtype is None.
+    """
+
+    def __init__(
+        self,
+        distance: Distance,
+        device: tp.Union[torch.device, str],
+        batch_size: int,
+        subjects_factors: tp.Union[np.ndarray, sparse.csr_matrix, torch.Tensor],
+        objects_factors: tp.Union[np.ndarray, torch.Tensor],
+        dtype: tp.Optional[torch.dtype] = torch.float32,
+    ):
+        assert batch_size > 0
+
+        self.dtype = dtype
+        self.device = torch.device(device)
+        self.batch_size = batch_size
+        self.distance = distance
+        self._scorer, self._higher_is_better = self._get_scorer(distance)
+
+        self.subjects_factors = self._normalize_tensor(subjects_factors)
+        self.objects_factors = self._normalize_tensor(objects_factors)
+
+    def rank(
+        self,
+        subject_ids: InternalIds,  # TODO check it
+        k: tp.Optional[int] = None,
+        filter_pairs_csr: tp.Optional[sparse.csr_matrix] = None,
+        sorted_object_whitelist: tp.Optional[InternalIdsArray] = None,
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+        """Rank objects to proceed inference using implicit library topk cpu method.
+
+        Parameters
+        ----------
+        subject_ids : csr_matrix | np.ndarray
+            Array of ids to recommend for.
+        k : Optional[int]
+            Derived number of recommendations for every subject id.
+            Return all recs if None.
+        filter_pairs_csr : sparse.csr_matrix, optional, default ``None``
+            Subject-object interactions that should be filtered from recommendations.
+            This is relevant for u2i case.
+        sorted_object_whitelist : sparse.csr_matrix, optional, default ``None``
+            Whitelist of object ids.
+            If given, only these items will be used for recommendations.
+            Otherwise all items from dataset will be used.
+
+        Returns
+        -------
+        (InternalIds, InternalIds, Scores)
+            Array of subject ids, array of recommended items, sorted by score descending and array of scores.
+        """
+        filter_viewed = filter_pairs_csr is not None
+        if filter_viewed and filter_pairs_csr.shape[0] != len(subject_ids):
+            explanation = "assumed that filter_pairs_csr and subject_ids are aligned"
+            raise ValueError(explanation)
+
+        if sorted_object_whitelist is None:
+            sorted_object_whitelist = np.arange(self.objects_factors.shape[0])
+
+        if not isinstance(subject_ids, np.ndarray):
+            subject_ids = np.array(subject_ids)
+
+        if k is None:
+            k = len(sorted_object_whitelist)
+
+        user_ids = subject_ids
+        sorted_item_ids_to_recommend = sorted_object_whitelist
```

Comment:
Could you please have these two not changed in namings?
It is easier to read code when names don't change for the same things.

And when names change it usually means that objects with new names should have some difference from objects with old names.